test/unit/test_sdb.c

```diff
@@ -207,6 +207,235 @@ bool test_sdb_copy() {
+	return db;
+}
+
+static int tmpfile_new (const char *filename, const char *buf, size_t sz) {
+	int fd = open (filename, O_RDWR | O_CREAT | O_TRUNC);
+	if (fd < 0) {
+		return -1;
+	}
+	if (!buf) {
+		return fd;
+	}
+	int w = write (fd, buf, sz);
+	if (w < (int)sz) {
+		close (fd);
+		return -1;
+	}
+	lseek (fd, 0, SEEK_SET);
+	return fd;
+}
+
+bool test_sdb_text_save_simple() {
+	Sdb *db = text_ref_simple_db ();
+
+	int fd = tmpfile_new (".text_save_simple", NULL, 0);
+	bool succ = sdb_text_save_fd (db, fd, true);
+	lseek (fd, 0, SEEK_SET);
+	char buf[0x1000];
+	memset (buf, 0, sizeof (buf));
+	read (fd, buf, sizeof (buf) - 1);
+	close (fd);
+	unlink (".text_save_simple");
+
+	sdb_free (db);
+
+	mu_assert_true (succ, "save success");
+	mu_assert_streq (buf, text_ref_simple, "text save");
+
+	mu_end;
+}
+
+bool test_sdb_text_save() {
+	Sdb *db = text_ref_db ();
+
+	int fd = tmpfile_new (".text_save", NULL, 0);
+	bool succ = sdb_text_save_fd (db, fd, true);
+	lseek (fd, 0, SEEK_SET);
+	char buf[0x1000];
+	memset (buf, 0, sizeof (buf));
+	read (fd, buf, sizeof (buf) - 1);
+	close (fd);
+	unlink (".text_save");
+
+	sdb_free (db);
+
+	mu_assert_true (succ, "save success");
+	mu_assert_streq (buf, text_ref, "text save");
+
+	mu_end;
+}
+
+static void diff_cb(const SdbDiff *diff, void *user) {
+	char buf[2048];
+	if (sdb_diff_format (buf, sizeof (buf), diff) < 0) {
+		return;
+	}
+	printf ("%s\n", buf);
+}
+
+bool test_sdb_text_load_simple() {
+	int fd = tmpfile_new (".text_load_simple", text_ref_simple, strlen (text_ref_simple));
+	Sdb *db = sdb_new0 ();
+	bool succ = sdb_text_load_fd (db, fd);
+	close (fd);
+	unlink (".text_load_simple");
+
+	mu_assert_true (succ, "load success");
+	Sdb *ref_db = text_ref_simple_db ();
+	bool eq = sdb_diff (ref_db, db, diff_cb, NULL);
+	sdb_free (ref_db);
+	sdb_free (db);
+	mu_assert_true (eq, "load correct");
+	mu_end;
+}
+
+bool test_sdb_text_load() {
+	int fd = tmpfile_new (".text_load", text_ref, strlen (text_ref));
+	Sdb *db = sdb_new0 ();
+	bool succ = sdb_text_load_fd (db, fd);
+	close (fd);
+	unlink (".text_load");
+
+	mu_assert_true (succ, "load success");
+	Sdb *ref_db = text_ref_db ();
+	bool eq = sdb_diff (ref_db, db, diff_cb, NULL);
+	sdb_free (ref_db);
+	sdb_free (db);
+	mu_assert_true (eq, "load correct");
+	mu_end;
+}
+
+bool test_sdb_text_load_bad_nl() {
+	int fd = tmpfile_new (".text_load_bad_nl", text_ref_bad_nl, strlen (text_ref_bad_nl));
+	Sdb *db = sdb_new0 ();
+	bool succ = sdb_text_load_fd (db, fd);
+	close (fd);
+	unlink (".text_load_bad_nl");
+
+	mu_assert_true (succ, "load success");
+	Sdb *ref_db = text_ref_bad_nl_db ();
+	bool eq = sdb_diff (ref_db, db, diff_cb, NULL);
+	sdb_free (ref_db);
+	sdb_free (db);
+	mu_assert_true (eq, "load correct");
+	mu_end;
+}
+
+bool test_sdb_text_load_broken() {
+	int fd = tmpfile_new (".text_load_broken", text_ref_broken, strlen (text_ref_broken));
+	Sdb *db = sdb_new0 ();
+	bool succ = sdb_text_load_fd (db, fd);
```

Comment:
imho sdb_load() should work for plain and bin databases. exposing text/binary APIs is not good. i would even prefer adding a parameter to the sdb_load api to specify the file format (TEXT, BINARY, GUESS). What do you think?

Reply:
You mean sdb_open?